src/datumaro/components/launcher.py

```diff
@@ -2,104 +2,121 @@
 #
 # SPDX-License-Identifier: MIT
 
-from typing import Generator, List
+import inspect
+import logging as log
+import os.path as osp
+from importlib.util import module_from_spec, spec_from_file_location
+from typing import List, Optional, Sequence, Tuple
 
 import numpy as np
 
-from datumaro.components.annotation import Annotation, AnnotationType, LabelCategories
+from datumaro.components.abstracts.model_interpreter import IModelInterpreter, ModelPred, PrepInfo
+from datumaro.components.annotation import Annotation
 from datumaro.components.cli_plugin import CliPlugin
-from datumaro.components.dataset_base import DatasetItem, IDataset
-from datumaro.components.transformer import Transform
-from datumaro.util import take_by
+from datumaro.components.dataset_base import DatasetItem
+from datumaro.components.media import Image
+from datumaro.errors import DatumaroError
 
 
 # pylint: disable=no-self-use
 class Launcher(CliPlugin):
-    def __init__(self, model_dir=None):
+    def __init__(self, model_dir: Optional[str] = None):
         pass
 
-    def launch(self, inputs):
+    def preprocess(self, img: Image) -> Tuple[np.ndarray, PrepInfo]:
+        """Preprocess single image before launch()
+
+        Datumaro passes image data as `np.ndarray` with BGR format (H, W, C).
+        The output should be also `np.ndarray` but it can be stacked into a batch of images.
+        In this step, you usually implement resizing, normalizing, or color channel conversion
+        for your launcher (or model).
+        """
         raise NotImplementedError()
 
+    def infer(self, inputs: np.ndarray) -> List[ModelPred]:
+        """
+        It executes the actual model inference for the inputs.
+
+        Parameters:
+            inputs: Batch of the numpy formatted input data
+        Returns:
+            List of model outputs
+        """
+        raise NotImplementedError()
+
+    def postprocess(self, pred: ModelPred, info: PrepInfo) -> List[Annotation]:
+        raise NotImplementedError()
+
+    def launch(self, batch: Sequence[DatasetItem]) -> List[List[Annotation]]:
+        """Launch to obtain the inference outputs of items.
+
+        Parameters:
+            inputs: batch of Datasetitems
+
+        Returns:
+            A list of annotation list. Each annotation list is mapped to the input
+            :class:`DatasetItem`. These annotation list are pseudo-labels obtained by
+            the model inference.
+        """
+        if len(batch) == 0:
+            return []
+
+        inputs_img = []
+        inputs_info = []
+
+        for item in batch:
+            prep_img, prep_info = self.preprocess(item.media)
+            inputs_img.append(prep_img)
+            inputs_info.append(prep_info)
+
+        preds = self.infer(
+            np.stack(
+                [np.atleast_3d(img) for img in inputs_img],
```

Comment:
`preprocess` has already defined to return batch of BGR formatted image. Should `np.atleast_3d` require?

Reply:
Removed in https://github.com/openvinotoolkit/datumaro/pull/1055/commits/429e2b13da67a8b1a48276749c3e1dd0077d81b1